code/FsComposite/Composite.Github/Request.fs

```diff
@@ -0,0 +1,141 @@
+        | _ -> ll (Request obj)
+
+    /// ... use request: add parameter "state": "open". ...
+    let rec pOpen obj =
+        match obj with
+        | Request (PrsReadRequest o) -> ll (Request (PrsReadRequest (o.AddParameter("state", "open"))))
+//        | RequestAllPages (o,t) ->  ll (RequestAllPages(pOpen o, t))
+        | _ -> ll obj
+
+    /// ... use request: add parameter "page": [number]. ...
+    let rec pPage number obj =
+        match obj with
+        | PrsReadRequest o -> ll (Request (PrsReadRequest (o.AddParameter("page", number))))
+        | _ -> ll (Request obj)
+
+    let rec allPages obj =
+        match obj with
+        | Request o -> ll (Request (RequestAllPages o))
+        | _ -> ll obj
+
+    let matchRequest getResponseMethod obj =
+        match obj with
+        | PrsReadRequest x -> PrsReadJson (x |> getResponseMethod)
+        | LabelsReadRequest x -> LabelsReadJson (x |> getResponseMethod)
+        | LabelsAttachRequest x -> LabelsAttachedJson (x |> getResponseMethod)
+        | LabelDettachRequest x -> LabelDettachedJson (x |> getResponseMethod)
+        | PrFilesReadRequest x -> PrFilesReadJson(x |> getResponseMethod)
+        | PrCommentsReadRequest x -> PrCommentsReadJson(x |> getResponseMethod)
+        | IssueCommentsReadRequest x -> IssueCommentsReadJson(x |> getResponseMethod)
+        | PrCommitsReadRequest x -> PrCommitsReadJson(x |> getResponseMethod)
+
+    /// ... use request : turn it into json using REST [client]...
+    let rec executeSingle (client : RestClient) obj =
+        let addJsonBody body (req : IRestRequest) =
+            req.RequestFormat <- DataFormat.Json
+            req.AddBody(body)
+        let getResponseContent req =
+            client.Execute(req).Content
+
+        let rec executeWith getResponseMethod obj2 =
+                try
+                    match obj2 with
+                    | Request (RequestSetInBody (o, s)) -> if s = Set.empty
+                                                           then Response (Message "Request was supposed to have a body. Execution cancelled.")
+                                                           else Request o |> executeWith (addJsonBody s >> getResponseMethod)
+                    | Request req -> Response (matchRequest getResponseMethod req)
+                with
+                | o -> Response (Error o)
+
+        obj |> executeWith getResponseContent
+
+    // function that will determine is json haven't any entities
+    let isEmptyResult obj =
+        match obj with
+        | Response (PrsReadJson x) -> if x.Length < 5 
+                                      then Some obj 
+                                      else None
+
+    let executeAllPages client obj =
+        let rec getResultFromPage startPageNumber results =
+            match obj with
+            | Request x ->
+                match pPage startPageNumber x with
+                | Cons(r, Nil) -> match isEmptyResult (executeSingle client r) with

```

Comment:
'is empty result' might not be the way we want to go. 
we don't actually want to know whether the result is an empty set of something in and of itself, but rather if there are more pages out there to load. obviously, for each json response type that would differ. in some cases we would be able to tell that by simply reading a single value of 'total pages' or 'next page' from the response. 
alternatively, we could treat the page json as a json array object with more json objects in it and use the 'is empty page' approach.